runtime/stdlib/account.go

```diff
@@ -2989,6 +3009,151 @@ func newPublicAccountCapabilitiesValue(
 	)
 }
 
+func getCheckedCapabilityController(
+	inter *interpreter.Interpreter,
+	capabilityAddressValue interpreter.AddressValue,
+	capabilityIDValue interpreter.UInt64Value,
+	wantedBorrowType *sema.ReferenceType,
+	capabilityBorrowType *sema.ReferenceType,
+) (
+	interpreter.CapabilityControllerValue,
+	*sema.ReferenceType,
+) {
+
+	if wantedBorrowType == nil {
+		wantedBorrowType = capabilityBorrowType
+	} else {
+		// Ensure requested borrow type is not more permissive
+		// than the capability's borrow type:
+		// The requested type must be a supertype
+
+		if !sema.IsSubType(capabilityBorrowType, wantedBorrowType) {
+			return nil, nil
+		}
+	}
+
+	capabilityAddress := capabilityAddressValue.ToAddress()
+	capabilityID := uint64(capabilityIDValue)
+
+	controller := getCapabilityController(inter, capabilityAddress, capabilityID)
+	if controller == nil {
+		return nil, nil
+	}
+
+	// Ensure requested borrow type is not more permissive
+	// than the controller's borrow type:
+	// The requested type must be a supertype
+
+	controllerBorrowStaticType := controller.CapabilityControllerBorrowType()
+
+	controllerBorrowType, ok :=
+		inter.MustConvertStaticToSemaType(controllerBorrowStaticType).(*sema.ReferenceType)
+	if !ok {
+		panic(errors.NewUnreachableError())
+	}
+
+	if !sema.IsSubType(controllerBorrowType, wantedBorrowType) {
+		return nil, nil
+	}
```

Comment:
What is the relationship between the type of a controller's borrow type and its controlled capability's borrow type. Do we guarantee that one is a subtype of the other? If so it should be possible to eliminate one of these two subtyping checks.

Reply:
The situation is unfortunately complicated and mainly stems from the behaviour of the `Capability` type/value.

A capability value may have itself a type, which determines how the capability should be borrowed. This is `capabilityBorrowType`. 
Optionally, the capability may then be borrowed with a different type, via `borrow<T>`. This is `wantedBorrowType`.
We first need to ensure that `wantedBorrowType` is not a subtype of `capabilityBorrowType`: https://github.com/onflow/cadence/pull/2476/files#diff-226624e4ec99d61a29cefaff93951c94ce45969551618dc912421100bf851c7bR3025.
In this check, the controller is not yet involved.

Then, here we have to check that the wanted borrow type is also not a subtype of the controller's borrow type, which determines how the controlled capability may get borrowed.